### ISSUE TITLE
yay -Syu --devel does not update devel packages

### ISSUE THREAD

**Issue by DrDoctor13** [Jguer/yay#902]
#### Affected Version
yay v9.1.0.r16.g808f633 - libalpm v11.0.2

#### Issue
Running `yay -Syu --devel` does not update -git packages, even after running `yay -Y --gendb`.

#### Output
```
➜  ~ yay -Syu --devel
[sudo] password for bucks: 
:: Synchronizing package databases...
 core is up to date
 extra is up to date
 community is up to date
 multilib is up to date
:: Searching databases for updates...
:: Searching AUR for updates...
:: Checking development packages...
:: Starting full system upgrade...
 there is nothing to do
```
The packages I am trying to update are latte-dock-git and oh-my-zsh-git. Both have had commits recently at the time of writing but yay does not detect them. Here are the two versions on my computer currently:
`oh-my-zsh-git r4895.c49486963-1`
`latte-dock-git 0.7.96.r514.gbc72c859-1`
Meanwhile, the current git hashes are `52afbf77f` and `cf0182e82`, respectively.

**Comment by parkerlreed** [comment]
Check your vcs.json. I found it wasn't actually adding some of them after the compile, hence it being unaware of updates. Wasn't able to track down the issue.

**Comment by Morganamilo** [comment]
Stiff has been a little broken on the master. No Idea if @parkerlreed is right about that issue. But as things are on the current master everything seems to be working now.

**Comment by parkerlreed** [comment]
I'm testing right now to see if I can reproduce the issue.

**Comment by DrDoctor13** [comment]
I've updated `yay` to the latest commit and the issue still persists. I ran `yay -S yay-git` and then `yay -Y --gendb` but the `vcs.json` file was not modified. In fact, it's completely missing `yay-git` and `latte-dock-git`. It does have `oh-my-zsh-git`, but the hash does not match either the currently installed version or the latest version.

As a point of order, I'm assuming that the correct file to be checking here is `~/.cache/yay/vcs.json`.

**Comment by DrDoctor13** [comment]
Here is the file. These hashes appear to be incorrect and way out of date.

```
{
	"konsole-dracula-git": {
		"github.com/dracula/konsole": {
			"protocols": [
				"git",
				"https"
			],
			"branch": "HEAD",
			"sha": "aa5d31e9d3480f407ec79adfea889c47fa2e8685"
		}
	},
	"materia-theme-git": {
		"github.com/nana-4/materia-theme.git": {
			"protocols": [
				"https"
			],
			"branch": "HEAD",
			"sha": "6386a791691e15ae0307836d134a05e91dd8ecbf"
		}
	},
	"obs-linuxbrowser": {
		"github.com/bazukas/obs-linuxbrowser.git": {
			"protocols": [
				"https"
			],
			"branch": "HEAD",
			"sha": "c4758561920f047a54a2ff145aaa3c89a0c561bd"
		}
	},
	"oh-my-zsh-git": {
		"github.com/robbyrussell/oh-my-zsh.git": {
			"protocols": [
				"git"
			],
			"branch": "HEAD",
			"sha": "fabee55948776e2e4c210e9dcd75e7bc38c02bec"
		}
	},
	"oxygen-gtk3-git": {
		"anongit.kde.org/oxygen-gtk": {
			"protocols": [
				"git"
			],
			"branch": "gtk3",
			"sha": "705b755f67b50213ec596a51b8cfabf2780bc24b"
		}
	}
}
```

**Comment by Morganamilo** [comment]
Running gendb pulls in the hashes of the current masters, regardless of what version you have installed. So after running gendb yay will then assume everything is up to date until a new commit appears.

But if they are indeed out of date and some are missing that is indeed a problem.

I'll need to look into is soon.

In the meantime if you delete your vcs.json and then gendb is it any different?

**Comment by DrDoctor13** [comment]
If I delete `vcs.json` from `~/.cache/yay` and gendb, a new file is not generated.

**Comment by Morganamilo** [comment]
Weird. Have you set any custom XDG environment variables?.

**Comment by DrDoctor13** [comment]
None to my knowledge. How can I check?

**Comment by Morganamilo** [comment]
The enviroment variables are documented in the manpage. Although if you have not set them that should not be an issue. Maybe you an trace yay and see what files it tries to access.

**Comment by DrDoctor13** [comment]
From what I can tell, it does try to access vcs.json, but says there's no file and continues. I've attached the output from strace.
[yaybug.txt](https://github.com/Jguer/yay/files/2933627/yaybug.txt)

**Comment by DrDoctor13** [comment]
For what it's worth: Tested in a fresh install of Manjaro with yay v9.1.0.r27.gec5746f and libalpm v11.0.3. Installed yay-git with pamac then ran `yay --gendb` and vcs.json was still not generated in ~/.cache/yay.

**Comment by Morganamilo** [comment]
https://github.com/Jguer/yay/blob/5fe9160b1bc1aa2bd53e2bc8c08959e4a819f0c0/vcs.go#L228

This is the line were it opens the file. It opens it with `os.O_RDWR|os.O_CREATE|os.O_TRUNC` but the trace says `O_RDONLY|O_CLOEXEC`. No idea why it's different.

**Comment by Jguer** [comment]
@DrDoctor13 Are you using gcc-go by any chance?

**Comment by DrDoctor13** [comment]
No @Jguer

**Comment by severinkaderli** [comment]
I have the same problem. Running `yay -Syu --devel` does not update any packages even if there are updates available. I tried with the stable version and with yay-git.

**Comment by Jguer** [comment]
Ok, I'll look into this this week

**Comment by DrDoctor13** [comment]
Tested yay v9.1.0.r43.ga75506c - libalpm v11.0.3, vcs.json is still not generated after running on my Manjaro installation. Will test on a clean Manjaro installation later.

**Comment by soredake** [comment]
Same problem on archlinux with yay-git, vcs.json is present but no devel packages updated.

**Comment by Morganamilo** [comment]
Still works just fine for me. I think the one way to really make any progress is for someone bisect this or at least test against an older version.

**Comment by Jguer** [comment]
Yes, I have been unable to reproduce it on my end

**Comment by severinkaderli** [comment]
It suddenly works for me again...

**Comment by Morganamilo** [comment]
Is that a bad thing or not? I'm not sure.

**Comment by DrDoctor13** [comment]
yay -Y --gendb still does not do anything for me, tested on a clean Manjaro installation

**Comment by Morganamilo** [comment]
Can you test it on an arch install at least? Not really sure why there would be any difference though.

**Comment by DrDoctor13** [comment]
Tested on a clean Arch install and it seemed to work fine. One point I noticed was that yay-git in the Arch install will report finding git repos after installing git packages, but not on my main Manjaro install. After installing new -git packages on Manjaro and running yay -Y --gendb, a vcs.json file was generated, but it was empty. It did not seem to find the git repos.

**Comment by Morganamilo** [comment]
Does manjaro ship a weird .gitconfig or something?

Anything non standard?

Does `git ls-remote https://github.com/Jguer/yay` work from the command line?

Do earlier versions have this problem? Say `v7.887`?

**Comment by DrDoctor13** [comment]
My .gitconfig seems fairly normal to me:

```
[user]
	email = [redacted]
	name = [redacted]
[core]
	autocrlf = input
```
Running `git ls-remote` works. Trying gendb with yay 7.887 does not.

**Comment by HarleyLorenzo** [comment]
I've been having these same issues on my arch install with yay 9.4.7. I have been able to get it to _sometimes_ work by bruteforcing gendb after a devel package install to force it to actually go into vcs.json (which would typically be void of the new devel package), but it takes several tries. 

And with devel package updates, I have to run -Syu several times (in another bruteforce manner) before it finally picks up on the remote changes, and actually updates. 

This both of my real hardware arch installs, and every arch VM I've made.

I'm happy to do whatever to help fix this bug.
